ProcessList.c

```diff
@@ -128,7 +137,131 @@ int ProcessList_size(ProcessList* this) {
+   //
+   // In the below example note how filtering a range of indices i is different from filtering for processes in the bounds left_bound < x < right_bound …
+   //
+   // The nested tree set is sorted by left value, which is guaranteed upon entry/exit of this function.
+   //
+   // i | l | r
+   // 1 | 1 | 9
+   // 2 | 2 | 8
+   // 3 | 4 | 5
+   // 4 | 6 | 7
+   for (unsigned int i = leftBound; i < rightBound; i++) {
+      Process* proc = (Process*)Hashtable_get(this->displayTreeSet, i);
+      if (proc->tree_depth == deep && proc->tree_left > left && proc->tree_right < right) {
+         if (Vector_size(layer) > 0) {
+            Process* previous_process = (Process*)Vector_get(layer, Vector_size(layer)-1);
+            // Make a 'right_bound' of previous_process in a layer a current's process index.
+            //
+            // Use 'tree_depth' as a temporal variable.
+            // it is save to do as later 'tree_depth' will be renovated.
+            previous_process->tree_depth = proc->tree_index;
+         }
+         Vector_add(layer, proc);
+      }
+   }
+
+   // The loop above changes process-1 so the last process on the layer
+   // isn't updated by the that loop.
+   //
+   // Thus, if present, set the `rightBound` to the last process on the layer
+   if (Vector_size(layer) > 0) {
+      Process* previous_process = (Process*)Vector_get(layer, Vector_size(layer)-1);
+      previous_process->tree_depth = rightBound;
+   }
+
+   Vector_quickSort(layer);
+
+   int size = Vector_size(layer);
+   for (int i = 0; i < size; i++) {
+      Process* proc = (Process*) (Vector_get(layer, i));
+
+      unsigned int idx = (*index)++;
+      int newLeft = (*treeIndex)++;
+
+      int level = deep-1;
+      int currentIndent = indent == -1 ? 0 : indent | (1 << level);
+      int nextIndent = indent == -1 ? 0 : (i < size - 1) ? currentIndent : indent;
+
+      unsigned int newLeftBound = proc->tree_index;
+      unsigned int newRightBound = proc->tree_depth;
+      ProcessList_updateTreeSetLayer(this, newLeftBound, newRightBound, deep+1, proc->tree_left, proc->tree_right, index, treeIndex, nextIndent);
+
+      int newRight = (*treeIndex)++;
+
+      proc->tree_left = newLeft;
+      proc->tree_right = newRight;
+      proc->tree_index = idx;
+      proc->tree_depth = deep;
+
+      if (indent == -1) {
+         proc->indent = 0;
+      } else if (i == size - 1) {
+         proc->indent = -currentIndent;
+      } else {
+         proc->indent = currentIndent;
+      }
+
+      Hashtable_put(this->draftingTreeSet, proc->tree_index, proc);
+   }
+
+   Vector_delete(layer);
+}
+
+static void ProcessList_updateTreeSet(ProcessList* this) {
+   unsigned int index = 0;
+   unsigned int tree_index = 1;
+   int vsize = Vector_size(this->processes);
+
+   ProcessList_updateTreeSetLayer(this, 0, vsize, 0, 0, vsize*2+1, &index, &tree_index, -1);
+   assert((int)Hashtable_count(this->draftingTreeSet) == vsize);
```

Comment:
Why has the other post-exec assertion on `this->displayTreeSet == 0` been removed? Where and why is it violated?

I know that your previously used assertion on `Hashtable_count(this->displayTreeSet) == vsize` does not hold for new/dead processes since last scan, but at least the `this->draftingTreeSet` should be empty before the ProcessList_updateTreeSetLayer` call; or am I missing something?

Reply:
> Why has the other post-exec assertion on this->displayTreeSet == 0 been removed? Where and why is it violated?

As I've said.

> When I was fixing it I've noticed that for some reason which I didn't know about this-processes list updated somewhere but no in ProcessList_scan
So I deleted some debug lines which is affected by this fact.

But I am not sure where.

> I know that your previously used assertion on Hashtable_count(this->displayTreeSet) == vsize does not hold for new/dead processes since last scan, but at least the this->draftingTreeSet should be empty before the ProcessList_updateTreeSetLayer` call; or am I missing something?

Correct I am take it correctly.

It was removed by the same reason as above.

Essentially  displayTreeSet  `sometimes` is not equal to 0 after calling updateTreeSetLayer  as well as `sometimes` `Hashtable_count(this->displayTreeSet) == vsize` isn't true because on in the beggining of function `Vector_size(this->processes) != Hashtable_count(this->displayTreeSet)` which is I am not sure why.

Reply:
BTW ` this->displayTreeSet == 0` shouldn't related to that. I think.

Reply:
I've found the issue `ProcessList_buildTreeBranch` seems sometimes addes copies on initial build

Reply:
I seems to be fixed

Reply:
All assertions in this context restored and checked with the code.